### PR TITLE
browsr: 1.19.0 -> 1.21.0, mupdf: 1.23.6 -> 1.24.8, python312Packages.pymupdf: 1.23.6 -> 1.24.8, python312Packages.textual-universal-directory: 1.1.0 -> 1.5.0

### LINKED ISSUE #332384: python3*Packages.pymupdf: Runtime import failure in buildEnv
### Describe the bug

When using nixpkgs stable (nixos-24.05) the packages `python312Packages.pymupdf` and `python311Packages.pymupdf` both are not working properly in a nix shell env.

### Steps To Reproduce
```nix
# flake.nix
{
  description = "test pymupdf devshell";
  inputs = {
    nixpkgs.url = "github:nixos/nixpkgs?ref=nixos-24.05";
  };
  outputs =
    { self, nixpkgs, ... }:
    let
      system = "x86_64-linux";
      pkgs = import nixpkgs { inherit system; };
    in
    {
      devShells.${system}.default = pkgs.mkShell {
        packages = [
          (pkgs.python312.withPackages (pp: with pp; [ pymupdf ]))
        ];
      };
    };
}
```

Now check in the devshell

```
$ nix develop
$ python
>>> import fitz
Traceback (most recent call last):
  File "/nix/store/qv31saxij43ncc1r2891ck4vnjfav9jz-python3-3.12.4-env/lib/python3.12/site-packages/fitz/__init__.py", line 140, in <module>
    from . import mupdf
ImportError: cannot import name 'mupdf' from partially initialized module 'fitz' (most likely due to a circular import) (/nix/store/qv31saxij43ncc1r2891ck4vnjfav9jz-python3-3.12.4-env/lib/python3.12/site-packages/fitz/__init__.py)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/nix/store/qv31saxij43ncc1r2891ck4vnjfav9jz-python3-3.12.4-env/lib/python3.12/site-packages/fitz/__init__.py", line 142, in <module>
    import mupdf
ModuleNotFoundError: No module named 'mupdf'
```

### Expected behavior
pymupdf gets imported properly since this is approved in nixos stable.

### Additional context
- Exact issue occurs with `python311` as well.
- *Unrelated* but it is a build failure on unstable #331934

### Notify maintainers
@teto

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 6.6.43, NixOS, 24.11 (Vicuna), 24.11.20240804.5d97d39`
 - multi-user?: `yes`
 - sandbox: `relaxed`
 - version: `nix-env (Nix) 2.18.5`
 - channels(root): `"nixos"`
 - nixpkgs: `/nix/store/x2lvdghyrpvkifz0k910garzbd3wpjdg-source`
```

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by greg-hellings**
## Description of changes

A change in a previous PR lifted a Python 3.12 limitation on building browsr. The existing version then failed to build due to being out of date. This PR updates the rest of the dependency tree to build in Python 3.12 and to reflect their latest versions.

Closes #274882
Closes #290651

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by teto** [comment]
Looks fine. Considering the number of rebuild, this should probably target staging though

**Comment by phanirithvij** [comment]
what about https://github.com/NixOS/nixpkgs/pull/298783 https://github.com/NixOS/nixpkgs/pull/274882 https://github.com/NixOS/nixpkgs/pull/290651 etc. can they all be closed?

**Comment by greg-hellings** [comment]
> what about #298783 #274882 #290651 etc. can they all be closed?

It looks like they would be able to be closed if this gets merged

**Comment by phanirithvij** [comment]
Could you update the target branch for the pr to staging from master?

**Comment by greg-hellings** [comment]
> Could you update the target branch for the pr to staging from master?

Done

File: pkgs/applications/misc/mupdf/default.nix [hunk 1]
```diff
@@ -60,16 +60,15 @@ let
...
     url = "https://mupdf.com/downloads/archive/${pname}-${version}-source.tar.gz";
-    sha256 = "sha256-rBHrhZ3UBEiOUVPNyWUbtDQeW6r007Pyfir8gvmq3Ck=";
+    sha256 = "sha256-pRjZvpds2yAG1FOC1/+xubjWS8P9PLc8picNdS+n9Eg=";
   };
 
-  patches = [ ./0001-Use-command-v-in-favor-of-which.patch
-              ./0002-Add-Darwin-deps.patch
+  patches = [ ./0002-Add-Darwin-deps.patch
               ./0003-Fix-cpp-build.patch
             ];
```
Comment (SuperSandro2000) [inline pkgs/applications/misc/mupdf/default.nix:73]:
```suggestion
  patches = [
    ./0002-Add-Darwin-deps.patch
    ./0003-Fix-cpp-build.patch
  ];
```

**Review by SuperSandro2000 — COMMENTED** [review]
Can you also pick [`f9eb925` (#298783)](https://github.com/NixOS/nixpkgs/pull/298783/commits/f9eb9258cafd8359292be9c413a58590ee291dc7)

Also there are some additional changes in [`b476b96` (#298783)](https://github.com/NixOS/nixpkgs/pull/298783/commits/b476b9681d4d3ca3e29c3a8b83b3c22552bc8b40) and [`4a3db5a` (#298783)](https://github.com/NixOS/nixpkgs/pull/298783/commits/4a3db5a7999aff6fb1dfcd9c33a5cfb64dd8770e) which I think are worthwhile to carry over to here

**Review by thiagokokada — APPROVED** [review]

**Comment by thiagokokada** [comment]
> Can you also pick [`f9eb925` (#298783)](https://github.com/NixOS/nixpkgs/pull/298783/commits/f9eb9258cafd8359292be9c413a58590ee291dc7)
> 
> Also there are some additional changes in [`b476b96` (#298783)](https://github.com/NixOS/nixpkgs/pull/298783/commits/b476b9681d4d3ca3e29c3a8b83b3c22552bc8b40) and [`4a3db5a` (#298783)](https://github.com/NixOS/nixpkgs/pull/298783/commits/4a3db5a7999aff6fb1dfcd9c33a5cfb64dd8770e) which I think are worthwhile to carry over to here

I went and merged this since while those changes are nice to have, they're not essential, and `mupdf` is broken for a long time in `nixpkgs` so it is better to fix it and improve the package later.

**Comment by dotlambda** [comment]
> I went and merged this since while those changes are nice to have, they're not essential, and `mupdf` is broken for a long time in `nixpkgs` so it is better to fix it and improve the package later.

Then why were the other PRs closed??

**Comment by thiagokokada** [comment]
> > I went and merged this since while those changes are nice to have, they're not essential, and `mupdf` is broken for a long time in `nixpkgs` so it is better to fix it and improve the package later.
> 
> Then why were the other PRs closed??

Because the main thing that is the bump is done, the only changes are smaller ones that can be done targeting either master or staging.

I think at that point it will be better to just remake the PR since there are lots of conflicts.

**Comment by dotlambda** [comment]
> I think at that point it will be better to just remake the PR since there are lots of conflicts.

They can be closed once that's done.
This is unacceptable.

**Comment by thiagokokada** [comment]
> > I think at that point it will be better to just remake the PR since there are lots of conflicts.
> 
> They can be closed once that's done. This is unacceptable.

Feel free to reopen the PRs if you think I did a mistake. I only closed them to help reduce the huge amount of PRs we have.

No need to feel attacked.

**Comment by SuperSandro2000** [comment]
> and `mupdf` is broken for a long time in `nixpkgs` so it is better to fix it and improve the package later.

If it is broken for a longer time and there is no urgency, we can also take the extra day or two to take the improvements. 

Also apparently this PR was so rushed, that we immediately got a regression in #334596

**Comment by dotlambda** [comment]
> Also apparently this PR was so rushed, that we immediately got a regression in #334596

That's unrelated because this PR was merged into master.

**Comment by thiagokokada** [comment]
> Also apparently this PR was so rushed, that we immediately got a regression in #334596

Completely unrelated, this PR is targetting `staging`.

**Comment by thiagokokada** [comment]
> If it is broken for a longer time and there is no urgency, we can also take the extra day or two to take the improvements.

It is my opinion that sometimes we focus too much in fixing things that doesn't really help anyone. For example, the `meta.mainProgram` change, it is fine, but it is something so trivial that can be done in a separate PR targetting `master` directly. No need to go through `staging`, and no need to be backported to this PR.

IMO, this kind ends up just discouraging newcomers that end up doing good work fixing difficult packages like this one. But I am not sure if this is the correct place to discuss this issue either, feel free to ask me in Matrix or something if you want to discuss this issue further.

**Comment by SuperSandro2000** [comment]
If the change is so trivial IMO creating another PR is just creating overhead for reviewers and mergers.

**Comment by thiagokokada** [comment]
> If the change is so trivial IMO creating another PR is just creating overhead for reviewers and mergers. 

I think the overhead is way higher for the ones opening a PR than the ones reviewing it. And at least for low impact changes, I don't think it matters too much. Now, getting packages fixed has a much higher impact, IMO.

But we can agree to disagree here, not sure if it makes sense to comment much more on this particular issue. If I don't forget I will open a PR later applying the `meta.mainProgram` changes.

**Comment by greg-hellings** [comment]
> If the change is so trivial IMO creating another PR is just creating overhead for reviewers and mergers.

It also makes it an absolute PITB for someone who just wants to contribute a fix and isn't wanting to be nitpicked to death by drift in reviewer's preferences and optional fields since the package was first merged in. In this case, adding a mainProgram option would be trivial, but would require bouncing back through a git rebase that not every contributor would be comfortable with. The other two PRs linked make major changes to the mupdf package. If those changes are desired, they should be subject to a completely different process than just a quick build fix submission by someone who is not the package maintainer and who just needs the package to build in order to fix their own problem. I just wanted browsr to work, I didn't want to adopt mupdf. I just needed to fix mupdf in order to get there.

**Comment by vcunat** [comment]
FYI: 9b88d0593508

**Pull request merged**